File: src/platform/posix/fdscan.c
/* naive/bad implementation, just allocate a big set and throw in
 * the descriptors we could find that is used in the current
 * process */
#include <stdlib.h>
#include <unistd.h>
#include <poll.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/resource.h>

int arcan_fdscan(int** listout)
{
/* multiple caveats with this one, the implementation is not reliable -
 * 1. race conditions against rlimit being changed pre/post
 * 2. descriptors created and dup:ed high, then rlimit dropped are invisible
 */
	struct rlimit rlim;
	int lim = 512;
	if (0 == getrlimit(RLIMIT_NOFILE, &rlim))
		lim = rlim.rlim_cur;

	struct pollfd* set = malloc(sizeof(struct pollfd) * lim);
	if (!set)
		return -1;

	for (size_t i = 0; i < lim; i++)
		set[i] = (struct pollfd){
			.fd = i
		};

	if (-1 == poll(set, lim, 0)){
		free(set);
		return -1;
	}

	size_t count = 0;
	for (size_t i = 0; i < lim; i++){
		if (!(set[i].revents & POLLNVAL))
			count++;
	}

	int* buf;
	if (!count || (buf = malloc(sizeof(int) * count)) == NULL){
		free(set);
		return -1;
	}

/* always include stdin/stdout/stderr as someone deliberately closing
 * still means that libs etc. can assume or try to work with it */
	size_t pos = 0;
	for (size_t i = 0; i < lim && count; i++){
		if (!(set[i].revents & POLLNVAL) || i < 3)
			buf[pos++] = set[i].fd;
	}

	free(set);
	*listout = buf;

	return pos;
}

File: src/platform/posix/fdscan.zig
// Zig port of posix/fdscan.c
// Scans open file descriptors using poll() and returns a list of valid ones.

const std = @import("std");
const c = std.c;

export fn arcan_fdscan(listout: *?[*]c_int) c_int {
    var rlim: c.rlimit = undefined;
    var lim: usize = 512;
    if (c.getrlimit(.NOFILE, &rlim) == 0) {
        lim = @intCast(rlim.cur);
    }

    const set: [*]c.pollfd = @ptrCast(@alignCast(c.malloc(@sizeOf(c.pollfd) * lim) orelse return -1));
    defer c.free(set);

    for (0..lim) |i| {
        set[i] = .{ .fd = @intCast(i), .events = 0, .revents = 0 };
    }

    if (c.poll(set, @intCast(lim), 0) == -1) {
        return -1;
    }

    var count: usize = 0;
    for (0..lim) |i| {
        if (set[i].revents & std.os.linux.POLL.NVAL == 0) {
            count += 1;
        }
    }

    if (count == 0) return -1;
    const buf: [*]c_int = @ptrCast(@alignCast(c.malloc(@sizeOf(c_int) * count) orelse return -1));

    var pos: usize = 0;
    for (0..lim) |i| {
        if (pos >= count) break;
        if (set[i].revents & std.os.linux.POLL.NVAL == 0 or i < 3) {
            buf[pos] = set[i].fd;
            pos += 1;
        }
    }

    listout.* = buf;
    return @intCast(pos);
}
